Configurable active user series

pstibrany opened the pull request:
**What this PR does**: This PR makes tracking of active series in the blocks storage configurable (there are two parameters: how often to purge old series from memory and update metrics, and how old must series be to be considered inactive).

This makes it possible to track active series more precisely even if they are still in the TSDB head, but haven't been appended any samples recently.

In the current state of the PR, this is only wired into the blocks engine, and not chunks engine.

**Checklist**
- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

pracucci commented on pkg/ingester/ingester.go:
```diff
@@ -103,6 +105,8 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
 	f.DurationVar(&cfg.MetadataRetainPeriod, "ingester.metadata-retain-period", 10*time.Minute, "Period at which metadata we have not seen will remain in memory before being deleted.")
 
 	f.DurationVar(&cfg.RateUpdatePeriod, "ingester.rate-update-period", 15*time.Second, "Period with which to update the per-user ingestion rates.")
+	f.DurationVar(&cfg.ActiveSeriesUpdatePeriod, "ingester.active-series-update-period", 5*time.Minute, "How often to update active series metrics (blocks engine only).")
```
I would default to 1m, to have more accurate tracking (by default).

pracucci commented on pkg/ingester/ingester.go:
```diff
@@ -103,6 +105,8 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
 
 	f.DurationVar(&cfg.RateUpdatePeriod, "ingester.rate-update-period", 15*time.Second, "Period with which to update the per-user ingestion rates.")
+	f.DurationVar(&cfg.ActiveSeriesUpdatePeriod, "ingester.active-series-update-period", 5*time.Minute, "How often to update active series metrics (blocks engine only).")
+	f.DurationVar(&cfg.ActiveSeriesIdle, "ingester.active-series-idle", 15*time.Minute, "After what time is series considered to be inactive (blocks engine only).")
```
Maybe we should call the CLI flag / YAML option `active-series-idle-time` or `active-series-idle-timeout`?

pracucci replied:
```suggestion
	f.DurationVar(&cfg.ActiveSeriesIdle, "ingester.active-series-idle", 15*time.Minute, "After what time a series is considered to be inactive (blocks engine only).")
```

pracucci commented on pkg/storage/tsdb/active_series.go:
```diff
@@ -0,0 +1,168 @@
+
+// activeSeriesStripe holds a subset of the series timestamps for a single tenant.
+type activeSeriesStripe struct {
+	refsMu        sync.Mutex
```
It's called `refsMu` but doesn't project `ref` only. I would rename for clarity.

pracucci commented on pkg/storage/tsdb/active_series.go:
```diff
@@ -0,0 +1,168 @@
+	s.refsMu.Lock()
+	defer s.refsMu.Unlock()
+
+	nanos := now.UnixNano()
```
Renaming `nanos` to `nowNanos` could help read the code.

pracucci commented on pkg/ingester/ingester_v2.go:
```diff
@@ -364,6 +376,19 @@ func (i *Ingester) updateLoop(ctx context.Context) error {
+			continue
+		}
+
+		now := time.Now()
```
I would move this before the loop.

pracucci reviewed (commented):
Thanks for working on it! I have few concerns:
1. The new metric adds more cognitive load to users when trying to understand the differences between chunks and blocks storage. Don't have an answer, but if it could work for the chunks storage too would probably easier to understand.
2. I'm a bit scared of the lock contention in ActiveSeries (because it's an exclusive lock). I'm wondering if we switch the entry timestamp to an atomic, if a RWMutex could be used (and most of the time we would acquire the read because we expect the series to be already in the map).

pstibrany commented:
> Thanks for working on it! I have few concerns:
> 
> 1. The new metric adds more cognitive load to users when trying to understand the differences between chunks and blocks storage. Don't have an answer, but if it could work for the chunks storage too would probably easier to understand.

I agree. Let's do that. (Reason why I kept chunks out of this PR is basically to first see how well it works in practice.)

> 2. I'm a bit scared of the lock contention in ActiveSeries (because it's an exclusive lock). I'm wondering if we switch the entry timestamp to an atomic, if a RWMutex could be used (and most of the time we would acquire the read because we expect the series to be already in the map).

I will write a benchmark to see throughput of both approaches.

pstibrany commented:
Benchmark showing improvements when using RW lock and atomics to update timestamps.

```
name                                  old time/op    new time/op    delta
ActiveSeriesTest_single_series/50-4      154ns ± 2%      76ns ± 1%  -50.52%  (p=0.008 n=5+5)
ActiveSeriesTest_single_series/100-4     156ns ± 1%      76ns ± 1%  -51.25%  (p=0.008 n=5+5)
ActiveSeriesTest_single_series/500-4     158ns ± 1%      76ns ± 2%  -52.01%  (p=0.008 n=5+5)

name                                  old alloc/op   new alloc/op   delta
ActiveSeriesTest_single_series/50-4      0.00B          0.00B          ~     (all equal)
ActiveSeriesTest_single_series/100-4     0.00B          0.00B          ~     (all equal)
ActiveSeriesTest_single_series/500-4     0.00B          0.00B          ~     (all equal)

name                                  old allocs/op  new allocs/op  delta
ActiveSeriesTest_single_series/50-4       0.00           0.00          ~     (all equal)
ActiveSeriesTest_single_series/100-4      0.00           0.00          ~     (all equal)
ActiveSeriesTest_single_series/500-4      0.00           0.00          ~     (all equal)
```

pracucci commented on pkg/ingester/active_series.go:
```diff
@@ -0,0 +1,198 @@
+func (s *activeSeriesStripe) updateSeriesTimestamp(now time.Time, series labels.Labels, fp model.Fingerprint, labelsCopy func(labels.Labels) labels.Labels) {
+	nowNanos := now.UnixNano()
+
+	e := s.findEntryForSeries(fp, series)
```
I think there's a race condition: we get a reference of an `activeSeriesEntry` and then we do update it. In the meanwhile, the purge could remove it, because the read lock is kept only during `findEntryForSeries()`. However, the next remote write push will add it back, right? If so, LGTM, but I would like your opinion too.

pstibrany replied:
Well spotted. Should be fixed now. (There was another problem... since atomic integer was not a pointer, but part of "entry", which was stored in a slice as a plain struct, without pointer, it could actually *move* in memory when slice grows. That's why atomic integers are now pointers).

gouthamve commented:
> The new metric adds more cognitive load to users when trying to understand the differences between chunks and blocks storage. Don't have an answer, but if it could work for the chunks storage too would probably easier to understand.

I'd be okay adding this to chunks store too. How much overhead will this add?

pstibrany commented:
> > The new metric adds more cognitive load to users when trying to understand the differences between chunks and blocks storage. Don't have an answer, but if it could work for the chunks storage too would probably easier to understand.
> 
> I'd be okay adding this to chunks store too. How much overhead will this add?

By my counting, each `activeSeriesEntry` needs 40 bytes on its own + copy of the labels. For 1M series that's at least 40 MB, and if we estimate labels to take 200 bytes, that would be additional 200 MB.

(In blocks storage, we try to reuse labels copy if added to the ref cache at the same time. In chunks doing the same is not easy.)

pracucci commented on CHANGELOG.md:
```diff
@@ -35,6 +35,7 @@
 * [ENHANCEMENT] When a tenant accesses the Alertmanager UI or its API, if we have valid `-alertmanager.configs.fallback` we'll use that to start the manager and avoid failing the request. #3073
 * [ENHANCEMENT] Add `DELETE api/v1/rules/{namespace}` to the Ruler. It allows all the rule groups of a namespace to be deleted. #3120
 * [ENHANCEMENT] Experimental Delete Series: Retry processing of Delete requests during failures. #2926
+* [ENHANCEMENT] Ingester: added new metric `cortex_ingester_active_series` to track active series more accurately. Also added two options to control how often this metric is updated (`-ingester.active-series-update-period`) and max idle time for series to be considered inactive (`-ingester.active-series-idle-timeout`). #3153
```
I've just merged a PR to cut the CHANGELOG because of the upcoming `1.4.0` release. Could you rebase `master` and move your CHANGELOG entry to the top, under `## master / unreleased`, please?

pracucci reviewed (approved):
Excellent job. I left a couple of nits. As discussed offline, I would add a `enabled` flag (default to false) to not let every Cortex user paying the cost of this tracking if not required (this is usually required by vendors billing systems).

pstibrany commented:
> Excellent job. I left a couple of nits. As discussed offline, I would add a `enabled` flag (default to false) to not let every Cortex user paying the cost of this tracking if not required (this is usually required by vendors billing systems).

Added `-ingester.active-series-enabled` flag (defaults to false), and switched fingerprinting to use xxHash 64-bit, which is faster than fnv64a (more visible for bigger labels, so benchmark now uses bigger label), but requires extra small allocation:

```
name                         old time/op    new time/op    delta
ActiveSeries_UpdateSeries-4    1.03µs ± 6%    0.74µs ±17%  -28.25%  (p=0.008 n=5+5)

name                         old alloc/op   new alloc/op   delta
ActiveSeries_UpdateSeries-4      206B ± 0%      253B ±11%  +23.01%  (p=0.008 n=5+5)

name                         old allocs/op  new allocs/op  delta
ActiveSeries_UpdateSeries-4      2.00 ± 0%      3.00 ± 0%  +50.00%  (p=0.008 n=5+5)
```

pstibrany commented:
I've also made sure that if `-ingester.active-series-enabled` is `false`, then new metric doesn't appear in the metrics at all. Previous version of the chunks-code actually exported 0 value for each user, which is useless. Unfortunately this made PR bigger :( (alternative would be to not register the metric into to metrics registry, but still create all in-memory structures).

pracucci commented on CHANGELOG.md:
```diff
@@ -2,6 +2,8 @@
 
 ## master / unreleased
 
+* [ENHANCEMENT] Ingester: added new metric `cortex_ingester_active_series` to track active series more accurately. Also added options to control how whether active series tracking is enabled (`-ingester.active-series-enabled`, defaults to false), and how often this metric is updated (`-ingester.active-series-update-period`) and max idle time for series to be considered inactive (`-ingester.active ...
```
```suggestion
* [ENHANCEMENT] Ingester: added new metric `cortex_ingester_active_series` to track active series more accurately. Also added options to control whether active series tracking is enabled (`-ingester.active-series-enabled`, defaults to false), and how often this metric is updated (`-ingester.active-series-update-period`) and max idle time for series to be considered inactive (`-ingester.active-series-idle-timeout`). #3153
```

pracucci commented on docs/configuration/config-file-reference.md:
```diff
@@ -613,6 +613,18 @@ lifecycler:
+
+# Enable tracking of active series.
+# CLI flag: -ingester.active-series-enabled
+[active_series_enabled: <boolean> | default = false]
```
Should we prefix all such config options with `active_series_metrics` to make it more clear?

pracucci commented on pkg/ingester/user_state.go:
```diff
@@ -30,6 +31,8 @@ type userStates struct {
 	cfg     Config
 	metrics *ingesterMetrics
+	// Is tracking of active series enabled?
+	activeSeriesEnabled bool
```
Food for thought: tor simplicity, what if we remove this and initialise `*ActiveSeries` (will be never filled) and we just don't register the gauge metric to the registerer?

pstibrany replied:
I turned that option down at first as I expected lot of tests calling `newIngesterMetrics`. Turns out it was fewer places, so I've changed code to do that instead.

pracucci commented on pkg/ingester/ingester_v2_test.go:
```diff
@@ -262,6 +315,9 @@ func TestIngester_v2Push(t *testing.T) {
 			// Create a mocked ingester
 			cfg := defaultIngesterTestConfig()
 			cfg.LifecyclerConfig.JoinAfter = 0
+			if testData.disableActiveSeries {
+				cfg.ActiveSeriesEnabled = false
+			}
```
```suggestion
			cfg.ActiveSeriesEnabled = !testData.disableActiveSeries
```

pracucci reviewed (approved):
Good job, still LGTM! Left final nits.

bboreham commented:
> In the current state of the PR, this is only wired into the blocks engine, and not chunks engine.

Is this still true?

pstibrany commented:
> > In the current state of the PR, this is only wired into the blocks engine, and not chunks engine.
> 
> Is this still true?

Not anymore. Also there is now a flag to enable this, which is disabled by default.
